### DISCUSSION TITLE
Line of sight

### DISCUSSION THREAD

**Discussion by joukop** [recastnavigation/recastnavigation#724] [Q&A]
I've been working on a server for an ancient defunct PS2 MMORPG. There's a big and seamless world so I've generated about 2 gigabytes of navmeshes that are loaded/unloaded dynamically by the server as players move around. They are used for NPC movement only - the client takes care of player collisions.

I don't have server-side LoS calculations yet for spell casting etc. Since I have the navmesh system working already, it would be very convenient, considering memory management and general simplicity, to have the required geometry for LoS calculations somehow embedded in the navmesh files.

I don't need full-fledged universal LoS and could probably live with some false positives (not being able to cast even though there is partial LoS). The game has no jumping or flying so we can rule that for the required LoS considerations, both ray endpoints are located at about player height above a walkable polygon.

I was thinking that maybe just attaching a number like "wall height" or even a flag like "considerable wall exists" to polygon edges could allow me to work out the math for a reasonable enough approximation. I could then adapt the Recast's existing raycast function to consider this additional data.

Any suggestions, is this a good idea and where/how in the navmesh building process could it be implemented? Thanks

**Comment by grahamboree** [comment]
Interesting!  This is a bit hard to do for the _general_ case, but depending on where you can cut corners or make assumptions WRT your game world it could be simplified.   The generalized solution would likely be retaining rasterized mesh geometry in some kind of compact hierarchical data structure (sparse octtrees maybe?) and use that for resolving LoS.  That's probably going to be quite a substantial bit of data (depending a lot on the rasterization settings obv.) if your navmeshes are already 2GB.

You mention storing "wall height" for poly edges which might be a way to store really coarse collision information.  Generally speaking a nav poly edge is either: the edge of a navmesh tile, the start of a slope, or the edge of an obstacle...unless you're explicitly marking certain terrain types as unwalkable (e.g. lava) in which case it could be the edge of an area of that terrain type that might not necessarily block LoS.

In the rasterization code, it's looking at the slope of input polygons using their surface normal to determine if they're walkable or not - if they're too steep for the agent to walk up.  You could probably take that information and use it to get a crude "wall filter".  

Ultimately though, LoS is probably really difficult to extract from a navmesh because there are a lot of factors at play.  e.g. a long straight line through the navmesh might not have LoS and be completely obscured because the navmesh spans the crest of a hill or something.  You'd really need to lean on some world terrain constraints (e.g. maybe it's a logically flat 2d world?) to make something like this work.
